\newcommand{\nocontentsline}[3]{}
\newcommand{\tocless}[2]{\bgroup\let\addcontentsline=\nocontentsline#1{#2}\egroup}
\fancyhf{}
\fancyhead[LE,RO]{\nouppercase{\thepage}}
\fancyhead[LO]{\sc \nouppercase{Appendices}}
\fancyhead[RE]{\sc \nouppercase{Appendices}}

\addtocontents{toc}{\protect\setcounter{tocdepth}{ 0}}
\chapter{Supplementary materials for Chapter 2}\label{app:npscarf}
%\addtocontents{toc}{\protect\setcounter{tocdepth}{ 0}}
%\addcontentsline{toc}{chapter}{Appendix \thechapter}

\newpage

\begin{figure}[!hpt]
\centering
\includegraphics[width=15.84cm]{images/suppFigure1.pdf}
\caption[Alignment of the \npscarf{}'s assembly for \kp{} ATCC BAA-2146 to its reference genomes]{Alignment of the \npscarf{}'s assembly for \kp{} ATCC BAA-2146 to its draft
reference genomes (GeneBank Accession GCA\_000364385.2). The five contigs were
in complete agreement with the chromosome and four plasmids.}
\label{SF:alignKp2146}
\end{figure}

\begin{figure}[!hpt]
\centering
\includegraphics[width=9.76cm]{images/suppFigure2.pdf}
\caption[Alignment of the \npscarf{}'s assembly for \kp{} ATCC 13883 to its reference genomes]{Alignment of the \npscarf{}'s assembly for \kp{} ATCC 13883 to its draft
reference genomes (GeneBank Accession GCA\_000742135.1). 
For display convenience, IDs of reference sequences are abbreviated by the last two digits
before the dot in the original accession code: 
Contigs 1 and 2 were aligned to the reference Scaffold 18 (KN046818.1). 
Contig 4 was aligned to two Scaffolds 19 (KN046819.1) and
22 (KN046822.1) and Contig 3 to Scaffolds 20 (KN046820.1) and 21 (KN046821.1).}
\label{SF:alignKp13883}
\end{figure}


\begin{figure}[!hpt]
\centering
\includegraphics[width=19cm]{images/suppFigure3.pdf}
\caption[Alignment of the  \npscarf{}'s assembly for \sce{} W303 to the reference
genome of the S288C strain]{Alignment of the  \npscarf{}'s assembly for \sce{} W303 to the reference genome of the S288C strain. Ten chromosomes (II, IV, V, VII, IX, X, XI, XIII, XV, XVI and the mitochondrion) were constructed into individual contigs, three chromosomes (I, III and VIII) were into two contigs each, and three chromosomes (VI, XII and XIV) were fused into two contigs because of misassembly.}
\label{SF:alignW303rt}
\end{figure}


\begin{figure}[!hpt]
\centering
\includegraphics[width=19cm]{images/suppFigure4.pdf}
\caption[Alignment of the Canu's assembly for \sce{} W303 to the reference
genome of the S288C strain]{Alignment of the Canu's assembly for \sce{} W303 to the reference genome of the S288C strain.  Chromosomes II and XVI were fused onto three
contigs.
}
\label{SF:alignW303Canu}
\end{figure}


\begin{figure}[!hpt]
\centering
\includegraphics[width=19cm]{images/suppFigure5.pdf}
\caption[Alignment of the miniasm's assembly for \sce{} W303 to the reference
genome of the S288C strain]{Alignment of the miniasm's assembly for \sce{} W303 to the reference genome of the S288C strain. Chromosomes V and XI were fused onto three contigs.}
\label{SF:alignW303Min}
\end{figure}

\clearpage

Supplementary Table~\ref{T:mem} shows the memory usage of the tools on
the datasets described in the paper. The scaffolders (SSPACE, LINK and
\npscarf{}) were run on the short read assemblies outputed by SPAdes. We hence
only present the memory footprint from running these tools only. The memory
requirement for each pipeline should be the \emph{maximum} between memory usage
of the tool and SPAdes. For NaS and Nanocorr, the error correction steps were
distributed across hundreds of jobs, each consumed a small memory footprint. We
report here only the memory usage from running  The values reported in the table
were from running Celera Assembler. Note that we ran Celera Assembler on
differing configurations, and we reported the here the memory usage of the
configuration resulting in the most complete assembly.
Memory reported for Canu and Miniasm was the \emph{maximum} among all tasks
in the pipeline (including Pilon and BWA-MEM) because each task can be run one
at a time; However, the reported memory usage for \npscarf{} was the \emph{sum} of 
memory for running \npscarf{} and BWA-MEM.
\begin{table}[!hpt]
\centering
\caption{Memory usage (Gb) of the different tools}
\label{T:mem}
\begin{tabular}{lrrrrrrrrrr}
\hline
\toprule
 & \textbf{Kp2146} & \textbf{Kp13883} & \textbf{E.~coli K12} & \textbf{ST H58} & \textbf{Sc W303}\\
\hline
SPAdes & 35.67 & 34.32 & 34.24 & 10.45 & 85.99 \\
 + SSPACE & 3.41 & 4.01 & 5.09 & 2.39 & 36.84 \\
 + LINK & 28.47 & 16.33 & 44.83 & 19.89 & 233.49 \\
 + \npscarf{}{} (rt) & 2.11 & 1.11 & 2.32 & 1.27 & 4.27\\
% + \nname{} (b) & 0.67 & 0.62 & 0.62 & 0.62 & 2.37 \\
NaS + CA & 8.02 & 8.10 & 9.21 & 8.23 & 83.60 \\
Nanocorr + CA & 3.76 & 3.99 & 6.91 & 1.57 & 159.91 \\
Canu + Pilon & - & 6.78 & 6.30 & - & 56.20 \\
Miniasm + Pilon & 2.99 & 6.08 & 6.04 & 1.97 & 89.51 \\
\hline
\end{tabular}
\end{table}

\addtocontents{toc}{\protect\setcounter{tocdepth}{ 0}}
\chapter{Supplementary materials for Chapter 3}\label{app:npbarcode}
%\addtocontents{toc}{\protect\setcounter{tocdepth}{ 0}}
%\addcontentsline{toc}{chapter}{Appendix \thechapter}
\newpage

\begin{figure}[!hp]
\includegraphics[width=\textwidth]{images/vntr1.png}  
\caption[PCR barcode MinION sequencing with \npbarcode{}]
{Another application of \npbarcode{} with GUI for ONT sequencing using PCR barcoding kit (3 libraries, Albacore base-caller).}
\label{supp_fig:npbarcode_pcr}
\end{figure}

\begin{table}[!htb]
\caption{Information for each of 8 samples used in the Native Barcoding sequencing protocol.}
\label{supp_tab:sample}
\centering
\begin{tabular}{llllr}
\hline
\toprule
\textbf{ID} & \textbf{Strain}          & \textbf{Publish ID} & \textbf{Barcode ID} & \textbf{Input (ng)} \\\hline
\rowcolor{Gray} 
GP\_023            & \emph{Streptococcus pneumoniae} &        ATCC 700677              & NB01                & 1,000                \\
GN\_092            & \emph{Klebsiella pneumoniae}    &         3\_GR\_13~\cite{Miranda2018}             & NB02                & 1,500                \\
\rowcolor{Gray} 
GN\_093            & \emph{Acinetobacter baumannii}  &          n/a            & NB03                & 1,500                \\
GN\_096            & \emph{Klebsiella pneumoniae}    &         5\_GR\_13~\cite{Miranda2018}              & NB04                & 935                 \\
\rowcolor{Gray} 
GN\_101            & \emph{Pseudomonas aeruginosa}   &          n/a            & NB05                & 1,500                \\
GN\_106            & \emph{Klebsiella pneumoniae}    &         11\_BR\_13~\cite{Miranda2018}              & NB06                & 1,500                \\
\rowcolor{Gray} 
GN\_132            & \emph{Klebsiella quasipneumoniae}    &         21\_GR\_13~\cite{Miranda2018}              & NB07                & 1,000                \\
GN\_133            & \emph{Klebsiella pneumoniae}    &         22\_GR\_12~\cite{Miranda2018}              & NB08                & 1,500       \\\hline        
\end{tabular}
\end{table}

% From the native sequencing, we also run Porechop and poreFUME, together with Metrichor's demultiplexer for comparison. It is worth noting that since Metrichor's demultiplexer only works on 2D sequencing reads for the time we conduct the experiment, it is  therefore necessary to limit the comparison within this set.

% Each demultiplexer's output is a set of read bins where each bin is supposed to correspond to one target library. To evaluate the accuracy of the classification, we align all reads back to the references by BWA-MEM. Theoretically to be the absolute ground truth, every single read must align to exact one reference. However, this is not the case due to nanopore sequencing error rate (causing no hit found) and the similarity between 8 bacterial samples in use (causing confusions because of multiple and/or chimeric hits). 

\begin{table}[!hp]
\centering
\caption[Pairwise comparison between samples in Native Barcode Sequencing]{Pairwise comparison (using MUMmer) between 8 samples in our Native Barcode Sequencing. Value in cell $(i,j)$ is percentage (bases) of genome $i$ aligned to genome $j$. Highly identical genomes and their figures are highlighted.} 
\label{supp_tab:dnadiff}
\begin{tabular}{lllllllll}
\hline
\toprule
& GP\_023 & \textbf{GN\_092} & GN\_093 & \textbf{GN\_096} & GN\_101 & \textbf{GN\_106} & \textbf{GN\_132} & \textbf{GN\_133} \\ \hline
\rowcolor{Gray} GP\_023 & 100 & 0.30 & 0.16 & 0.27 & 0.09 & 0.34 & 0.28 & 0.30 \\
\textbf{GN\_092} & 0.14 & 100  & 0.30 & \textbf{87.28} & 1.09 & \textbf{92.74} & \textbf{78.42} & \textbf{97.11} \\
\rowcolor{Gray} GN\_093 & 0.13 & 0.32  & 100 & 0.32 & 0.30 & 0.25 & 0.23 & 0.3  \\
\textbf{GN\_096} & 0.14 & \textbf{88.84}  & 0.31 & 100 & 1.27 & \textbf{88.10}  & \textbf{79.75}  & \textbf{86.24}  \\
\rowcolor{Gray} GN\_101 & 0.03 & 0.90 & 0.15 & 1.00 & 100 & 1.02 & 0.73  & 0.72 \\ \textbf{GN\_106} & 0.17  & \textbf{93.96} & 0.17 & \textbf{87.47} & 1.19  & 100 & \textbf{79.65} & \textbf{91.86}  \\
\rowcolor{Gray} \textbf{GN\_132} & 0.13 & \textbf{79.99}  & 0.16  & \textbf{79.94} & 0.84 & \textbf{80.50} & 100 & \textbf{79.66} \\
\textbf{GN\_133} & 0.15 & \textbf{99.99} & 0.26 & \textbf{87.31} & 0.93    & \textbf{93.47} & \textbf{80.50} & 100 \\ \hline
\end{tabular}
\end{table}

% \tab{tab:dnadiff} shows the intersection of those genomes in pairs. From the table, 5 out of 8 samples shares significant identical genome's content. Especially there are three bacteria namely GN\_092, GN\_106 and GN\_133 possessing more than 90\% identity thus rooting great possibility of confusing alignments. On the other hand, 3 out of 8 genomes are  significantly unique when only sharing around 1\% or less of common bases to another. In fact, the only gram negative species GP\_23 is the most diverged isolate amongst all.

% Despite of aforementioned issues, for the comparison purpose, we can still have a decent estimation on the accuracy of the binning algorithms based on alignment. We extract the best out of all possible alignments for a read and assign its origin as the best-hit reference. Results are shown in \fig{fig:comparison}. For better illustration, divergences between samples are reflected via corresponding color palette, e.g. GN\_092, GN\_106 and GN\_133 are filled with nearly identical colors due to their highly similarity.

\begin{figure}[!hpt]
\includegraphics[width=\textwidth]{images/alignment.png}
\caption[Comparison of ONT Native Barcode Sequencing demultiplexing accuracy]
{Comparison of ONT Native Barcode Sequencing (8 libraries) demultiplexing accuracy. The bars present the number of aligned reads to references. Each group of bars correspond to a demultiplex bin. Bars in a group, from left to right, are from (1) the total dataset (2) \npbarcode{} (3) Porechop (4) poreFUME demultiplex. GN\_092, GN\_106, GN\_133 together with GN\_096 and GN\_132 are very close thus being filled by similar colors (pink-red).}
\label{supp_fig:comparison}
\end{figure}

% In general, the performances of 4 demultiplexers are stably comparable, especially \npbarcode{}, Porechop and Metrichor. The number of correctly aligned reads from \npbarcode{} is slightly higher than the other two while poreFUME's figures are the lowest among all. 
% In comparison with the total number of reads aligned to the reference, the number of corresponding reads discovered by demultiplexers are considerable close except for the cases of GN\_132 and GN\_133. We suggest that this is due to artifact during library preparation and/or barcode ligation step that destroy the integrity of the barcode sequences and make them more difficult to be detected.

\begin{table}[!hpt]
\centering
\caption[Statistics for identification of GP\_023]{True Positive and True Negative rate on identification of Gram Negative species GP\_023. 

It can be seen that \npbarcode{} has the highest rate of true positive (sensitivity 88.26\%) and the lowest rate of true negative (specificity 99.19\%). 
The common behavior for all demultiplex algorithms is to takes higher/lower risk of wrong classification in exchange for the greater/lesser discovery rate possible. This ratio has to be managed in a proper way depending on different situations yet can be adjusted by parameter calibration. For this use case with default parameter set, \npbarcode, Porechop and Metrichor built-in demultiplex return comparable results while poreFUME is slightly more conservative.} 
\label{supp_tab:sensitivity}
\begin{tabular}{l|llll}
\hline
\toprule
& \textbf{npBarcode} & \textbf{Porechop} & \textbf{poreFUME} & \textbf{Metrichor} \\ \hline
True Positive & 2,924 & 2,812 & 2,515 & 2,875 \\ 
True Negative & 22,125 & 22,147 & 22,177 & 22,155 \\ 
False Positive & 181 & 159 & 129 & 151 \\ 
False Negative & 389 & 501 & 798 & 438 \\ \hline
\textbf{Sensitivity (\%)} & 88.26 & 84.88 & 75.91 & 86.78 \\ 
\textbf{Specificity (\%)} & 99.19 & 99.29 & 99.42 & 99.32 \\  \hline
\end{tabular}
\end{table}

\begin{lstlisting}[float,language=bash,caption={An example of \emph{script.sh} used for \npbarcode{}, assuming that all SPAdes output folders are located in the same directory as this script and have name containing the barcoded sample.},label=lst:script]
#!/bin/bash
dirname=`find . -maxdepth 1 -type d -name "*${1}*" -print -quit`
bwa index ${dirname}/contigs.fasta
bwa mem -t 16 -k11 -W20 -r10 -A1 -B1 -O1 -E1 -L0 -a -Y -K 10000 \
	${dirname}/contigs.fasta - 2> /dev/null | \
jsa.np.npscarf -realtime -read 100 -time 1 -b - \
	-seq ${dirname}/contigs.fasta -spadesDir ${dirname} \
	-prefix ${1} > ${1}.log 2>&1
\end{lstlisting}


\begin{table}[!hpt]
\centering
\scriptsize
\caption[Real-time emulation of time to detect resistance genes from DNA sequencing]{Real-time emulation of time to detect resistance genes from DNA sequencing. \textbf{Bold} represents genes or gene family detected in final assembly and \# displays more than three genes grouped within this family. Genes displayed in order of time detected. Brackets represent class of antibiotic this gene confers resistance which includes: A, aminoglycoside; B, beta-lactam; F; fosfomycin; Fu; fusidic acid, M, macrolide; P, phenicol; Q, quinolone; R, rifampicin; S, sulphonamide; T, tetracycline; Tr, trimethoprim; V, vancomycin.}
\label{supp_tab:emulate}
\begin{tabular}{|l|c|}
\hline
{\small Time (mins)} & {\small Resistance gene(s) detected}  \\ \hline \hline
\multicolumn{2}{|c|}{\small Isolate: 1\_GR\_13 Total run time: 1279 mins} \\ \hline \hline

10          & \begin{tabular}[c]{@{}c@{}}\textbf{oqxA} (Q), \textbf{dfrA1} (Tr), \textbf{dfrA14} (Tr), \textbf{dfrA23} (Tr), \textbf{blaVIM-27}\# (B), \textbf{sul1/3} (S), \textbf{strA} (A), \\ \textbf{aph(3’)-Ia/c} (A), strB (A), aadB (A), \textbf{mph(A} (M), \textbf{blaTEM-1B}\# (B), \textbf{oqxB} (Q), \textbf{rmtB2} (A)\end{tabular}             \\ \hline
30          & \begin{tabular}[c]{@{}c@{}}sul2 (S), \textbf{ARR-2/3/6} (R), \textbf{blaOXA-10}\# (B), \textbf{blaVEB-1}\# (B), \textbf{tet(G)} (T), \textbf{cmlA1} (P), floR (P), \\ \textbf{blaSHV-11}\# (B), \textbf{fosA} (F)\end{tabular}                                                                      \\ \hline
60          & ARR-3 (R), aac(6’)Ib\# (A), ARR-7 (R), aac(2’) (A), cml (P)                                                                                                                                                                          \\ \hline
120         & aac(3’)-IIIc (A)                                                                                                                                                                                                                     \\ \hline
300         & aac(3’)-IIIb (A), aac(6’)-Ic (A),                                                                                                                                                                                                    \\ \hline
600         & blaPAO (B), aph(3’)-IIb (A), aph(6)-Ic (A)                                                                                                                                                                                           \\ \hline
900         & catpC233 (P), blaOKP (B)                                                                                                                                                                                                             \\ \hline
1200        & \textbf{tet(A)} (T)                                                                                                                                                                                                                           \\ \hline \hline
\multicolumn{2}{|c|}{\small Isolate: 2\_GR\_12 Total run time: 2468 mins}                                                                                                                                                                                                             \\ \hline \hline
10          & \textbf{blaKPC-2}\# (B), \textbf{blaTEM-1A}\#(B), \textbf{aac6Ib/-cr}\# (A), \textbf{cmlA1} (P), \textbf{dfrA12} (Tr), \textbf{sul1/ 3} (S)                                                                                                                                                \\ \hline
30          & \textbf{rmtB2} (A), \textbf{oqxA} (Q), \textbf{dfrA14} (Tr), blaPAO (B)                                                                                                                                                                                         \\ \hline
60          & \begin{tabular}[c]{@{}c@{}}\textbf{oqxB} (Q), strA (A), strB (A), \textbf{aph(3’)-1a/c} (A), \textbf{sul2} (S), \textbf{blaSHV-11/12}\# (B), \textbf{mph(A)} (M), \\ \textbf{catA1} (P), \textbf{tet(G)} (T), \textbf{blaOXA-9}\# (B), \textbf{aadA1/2}\# (A), fusB (Fu), \textbf{aac(6’)/aph(2”)} (A)\end{tabular}            \\ \hline
120         & \textbf{blaOXA-10}\# (B), floR (P), \textbf{ARR-2/3/6} (R), aadB (A), \textbf{fosA} (F), \textbf{blaVEB-1}\# (B), cml (P), \textbf{dfrA23} (Tr)                                                                                                                                   \\ \hline
300         & aac(3’)-IIIc (A), catpC233 (P), aac(3’)-IIIb (A), ARR-3 (R)                                                                                                                                                                          \\ \hline
600         & \textbf{tet(A)} (T)                                                                                                                                                                                                                           \\ \hline
900         & aph(6)-Ic (A)                                                                                                                                                                                                                        \\ \hline
1200        & -                                                                                                                                                                                                                                    \\ \hline \hline
\multicolumn{2}{|c|}{\small Isolate: 16\_GR\_13 Total run time: 1277 mins}                                                                                                                                                                                                            \\ \hline \hline
10          & blaOXA-436 (B), \textbf{sul1}/3 (S), \textbf{dfrA12} (Tr), \textbf{fosA} (F), aadB (A), strA (A), \textbf{sul2} (S), strB (A), blaCTX-M-64 (B)                                                                                                                           \\ \hline
30          & \begin{tabular}[c]{@{}c@{}}\textbf{blaOXA-48}\# (B), \textbf{aph(3’)-1a/c} (A), \textbf{mph(A)} (M), \textbf{rmtB2} (A), floR (P), \textbf{blaCTX-M-15}\# (B), \\ \textbf{aac(6’)Ib-cr}\# (A), \textbf{blaOXA-1}\# (B), \textbf{oqxB} (Q), \textbf{oqxA} (Q), \textbf{blaTEM-1B}\# (B), \textbf{blaVEB-1}\# (B), \textbf{cmlA1} (P)\end{tabular} \\ \hline
60          & \textbf{tet(G)} (T), \textbf{blaOXA-10}\# (B), \textbf{blaSHV-11}\# (B), \textbf{aac(3’)-IIa}\# (A), \textbf{ARR-2}/3/6 (R)                                                                                                                                                       \\ \hline
120         & \textbf{aadA1/2}\# (A), fusB (Fu), ermT (M), str (A), rmtG (A), \textbf{aac(6’)}/aph(2”) (A)                                                                                                                                                           \\ \hline
300         & ARR-3 (R), catpC233 (P), cml (P), aph(6)-Ic (A), aac(3’)-IIIb (A)                                                                                                                                                                    \\ \hline
600         & vanR (V), dfrA14 (Tr)                                                                                                                                                                                                                \\ \hline
900         & blaPAO (B)                                                                                                                                                                                                                           \\ \hline
1200        & aph(3’)-IIb (A)                                                                                                                                                                                                                      \\ \hline \hline
\multicolumn{2}{|c|}{\small Isolate: 20\_GR\_12 Total run time: 1277 mins}                                                                                                                                                                                                            \\ \hline \hline
10          & \textbf{aac(6’)Ib/-cr}\# (A), \textbf{blaTEM-1A}\# (B), \textbf{dfrA14} (Tr), \textbf{sul2} (S), strB (A), \textbf{blaKPC-2}\# (B),\textbf{ tet(A)} (T)                                                                                                                                    \\ \hline
30          & \textbf{oqxA} (Q), \textbf{blaSHV-11/12}\# (B), \textbf{aph(3’)-Ia} (A), \textbf{fosA} (F), \textbf{blaOXA-9} (B), \textbf{oqxB} (Q), aph(6)-Ic (A)                                                                                                                                        \\ \hline
60          & -                                                                                                                                                                                                                                    \\ \hline
120         & aac(2’) (A), catpC233 (P), aac(3’)-IIIb (A)                                                                                                                                                                                          \\ \hline
300         & ermT (M), blaPAO (B),\textbf{ aac(6’)}/aph(2”) (A), catpC221 (P)                                                                                                                                                                              \\ \hline
600         & rmtf (A), ermG (M)                                                                                                                                                                                                                   \\ \hline
900         & aac(3’)-IIIc (A), aac(6’)-Ic (A)                                                                                                                                                                                                     \\ \hline
1200        & vatB (M), ARR-2/3/6 (R), aadB (A)                                                                                                                                                                                                    \\ \hline
\end{tabular}
\end{table}

\normalsize

\addtocontents{toc}{\protect\setcounter{tocdepth}{ 0}}
\chapter{Supplementary materials for Chapter 4}\label{app:npgraph}
%\addtocontents{toc}{\protect\setcounter{tocdepth}{ 0}}
%\addcontentsline{toc}{chapter}{Appendix \thechapter}
\newpage

\makeatletter

\newlength\oriarrayrulewidth  
\newcount\orilowpenalty
\newcommand\nobreakmidrule{%
 \noalign{\global\oriarrayrulewidth\arrayrulewidth\relax
          \global\orilowpenalty\@lowpenalty\relax  
          \global\@lowpenalty=\numexpr-10000\relax%
          \global\arrayrulewidth\lightrulewidth\relax}
 \hline
 \noalign{\global\@lowpenalty=\orilowpenalty\relax%
          \global\arrayrulewidth\oriarrayrulewidth\relax}}

\makeatother

\begin{longtable}[!hpt]{llcrrrrr}
\caption{Benchmarking \npgraph{} against \npscarf{} versions, $\mathtt{hybridSPAdes}$ and \unicycler{} hybrid assembler with the synthetic dataset.}
\label{supp_tab:synthetic_benchmark}\\
\toprule
    &       & Assembly &  & N50  & Mis- &  Mismatch & Indel \\
    & Method & size (bp) & \#Contigs  & (bp) & assemblies & (per $100$Kbp) & (per $100$Kbp) \\
    \hline  
\endfirsthead
\multicolumn{8}{c}%
{\tablename\ \thetable\ -- \textit{Continued from previous page}} \\
\hline
    &       & Assembly &  & N50  & Mis- &  Mismatch & Indel \\
    & Method & size (bp) & \#Contigs  & (bp) & assemblies & (per $100$Kbp) & (per $100$Kbp) \\
\hline
\endhead
\hline \multicolumn{8}{r}{\textit{Continued on next page}} \\
\endfoot
\hline
\endlastfoot

\rowcolor{Gray}
\multicolumn{8}{l}{random sequences no repeats} \\* %
\nobreakmidrule
\rowcolor{Gray}
& npScarf &  4110000 &  3  &  4000000  &  0  & 0.00  & 0.00\\*
\rowcolor{Gray}
& npScarf\_wag & 4109516  &  3  &  4000000  &  0  & 0.00  & 0.00\\*
\rowcolor{Gray}
& npGraph-bwa & 4110000  &  3  &  4000000  &  0  & 0.00  & 0.00\\*
\rowcolor{Gray}
& npGraph-mm2 & 4110000  &  3  &  4000000  &  0  & 0.00  & 0.00\\*
\rowcolor{Gray}
& hybridSPAdes & 4110231  &  3  & 4000077   &  0  & 0.00  &  0.07\\*
\rowcolor{Gray}
& Unicycler & 4110000  &  3  &  4000000  &  0  & 0.00  &  0.00\\
\hline
\multicolumn{8}{l}{random sequences some repeats} \\* %
\nobreakmidrule
& npScarf & 4109612  &  3  &  4001483  &  0  &  0.00 &  1.22\\*
& npScarf\_wag & 4108128  &  3 &  3999999  &  0  &  0.00 &  0.95\\*
& npGraph-bwa & 4110000  &  3  &  4000000  &  0 &  0.00 &  0.00\\*
& npGraph-mm2 &  4110000 &  3  &   4000000 & 0  & 0.00  &  0.00\\*
& hybridSPAdes & 4108283  &  3  &  4000077  &  0  & 0.02  & 0.05\\*
& Unicycler &  4110000 &  3 & 4000000  &  0 & 0.00  &  0.00\\
\hline
\rowcolor{Gray}
\multicolumn{8}{l}{random sequences many repeats} \\* %
\nobreakmidrule
\rowcolor{Gray}
& npScarf &  4251391 &  9  &  3952225  &  27  & 0.88  & 7.86\\*
\rowcolor{Gray}
& npScarf\_wag &  4554192 &  9  &  3999620  &  37  & 0.32  & 5.84\\*
\rowcolor{Gray}
& npGraph-bwa & 4110000  &  3  &  4000000  &  0  &  0.32 & 0.15\\*
\rowcolor{Gray}
& npGraph-mm2 &  4110000 &  3  &  4000000  &  0  & 0.32  & 0.15\\*
\rowcolor{Gray}
& hybridSPAdes & 4108646  &  3  &  4000077  &  0  &  0.68 &  0.17\\*
\rowcolor{Gray}
& Unicycler & 4110000  &  3  &  4000000  &  0  &  0.32  & 0.15  \\
\hline
% \pagebreak
% \toprule
\multicolumn{8}{l}{\emph{Acinetobacter} A1} \\* %
\nobreakmidrule
& npScarf & 3918192  &  3  &  3899455  &  4  & 3.15  & 0.77 \\*
& npScarf\_wag & 4188998  &  5  &  3286149  &  6  & 6.94  & 1.87 \\*
& npGraph-bwa & 3917160  &  2  &  3908429  &  1 & 19.80  & 2.25 \\*
& npGraph-mm2 & 3918048  &  2  &  3909317  &  4 & 21.42  & 1.51 \\*
& hybridSPAdes &  3886253 &  53  &  1403208  &  0  & 5.62  & 0.26\\*
& Unicycler & 3917745  &  2 & 3909014  & 0  & 2.50  &  0.13\\
\hline
\rowcolor{Gray}
\multicolumn{8}{l}{\emph{Acinetobacter} AB30} \\* %
\nobreakmidrule
\rowcolor{Gray}
& npScarf & 4594626  &  11  &  4299677  &  1  & 69.95  & 51.04\\*
\rowcolor{Gray}
& npScarf\_wag & -  &  -  &  -  & -   &  - & -\\*
\rowcolor{Gray}
& npGraph-bwa & 4317408  &  6  &  2766938  &  1  & 38.10  & 1.72\\*
\rowcolor{Gray}
& npGraph-mm2 & 4336804  &  1  &  4336804  & 0   &  23.16 & 1.55\\*
\rowcolor{Gray}
& hybridSPAdes & 4286627  &  49  &  3308039  &  0  &  4.60 &  0.44\\*
\rowcolor{Gray}
& Unicycler &  4333041 &  1  &  4333041  &  1  &  6.42 &  0.53\\
\hline
\multicolumn{8}{l}{\ec{} K12 MG1655} \\* %
\nobreakmidrule
& npScarf & 4647010  &  5  &  4609437  &  9  & 10.65  & 17.61 \\*
& npScarf\_wag &  4678785 &  5  &  4641364  &  0  & 8.38  &  2.31\\*
& npGraph-bwa & 4643368  &  1  &  4643368  & 0  &  9.37 & 0.75 \\*
& npGraph-mm2 & 4643265  &  1  &  4643265  &  0 &  10.43 & 1.25 \\*
& hybridSPAdes &  4641097 &  1  &  4641097  &  0  & 1.19  & 0.13\\*
& Unicycler & 4641650  &  1 &  4641650 &  0 & 3.43  & 0.26 \\
\hline
\rowcolor{Gray}
\multicolumn{8}{l}{\ec{} O25b H4-ST131} \\* %
\nobreakmidrule
\rowcolor{Gray}
& npScarf &  5353639 &  7  &  5087544  &  14  &  22.43 & 6.69\\*
\rowcolor{Gray}
& npScarf\_wag & 5413499  &  7  &  5108146  &  6  &  26.34 & 3.95\\*
\rowcolor{Gray}
& npGraph-bwa & 5251882  &  3  &  5112329  &  0  &  15.26 & 1.11\\*
\rowcolor{Gray}
& npGraph-mm2 &  5250391 &  3  &  5110666  &  0  &  13.56 & 1.05\\*
\rowcolor{Gray}
& hybridSPAdes & 5249315  &  7  &  5109649  &  0  &  2.23 &  0.42\\*
\rowcolor{Gray}
& Unicycler & 5249442  &  3  &  5109760  &  0  & 4.02 & 0.27 \\
\hline
% \pagebreak
% \toprule
\multicolumn{8}{l}{\emph{Klebsiella} 30660 NJST258 1} \\* %
\nobreakmidrule
& npScarf & 5739591  &  9  &  5257627  &  8  & 11.55  & 12.93\\*
& npScarf\_wag & 5527526  &  5  &  5264334  &  0  & 9.86  & 2.43\\*
& npGraph-bwa & 5535507  &  8  &  5264082  &  2  & 12.77  & 1.50\\*
& npGraph-mm2 & 5534879  &  8  &  5263608  &  0  & 6.49  & 1.14\\*
& hybridSPAdes &  5528124 & 8   &  5263320  &  0  & 1.36  & 0.80 \\*
& Unicycler & 5537860  &  9  &  5263196  &  0  & 1.34  & 0.51 \\
\hline
\rowcolor{Gray}
\multicolumn{8}{l}{\emph{Klebsiella} MGH 78578} \\* %
\nobreakmidrule
\rowcolor{Gray}
& npScarf &  5676644 &  6  &  5308928  &  17  & 20.26  & 14.74\\*
\rowcolor{Gray}
& npScarf\_wag & 5672565  &  5  &  5315270  &  6  &  19.15 & 3.73\\*
\rowcolor{Gray}
& npGraph-bwa &  5698292 &  6  &  5316232  &  1  & 23.94  & 2.34\\*
\rowcolor{Gray}
& npGraph-mm2 & 5696273  &  6  &  5315757  &  0  &  19.35 & 1.84\\*
\rowcolor{Gray}
& hybridSPAdes &  5665442 &  19  &  5315086  &  0  & 5.28  &  0.92\\*
\rowcolor{Gray}
& Unicycler &  5694231 &  14  &  5315096  &  0  & 5.38 &  0.21\\
\hline
\multicolumn{8}{l}{\emph{Klebsiella} NTUH-K2044} \\* %
\nobreakmidrule
& npScarf & 5468717  &  3  &  5239437  &  5  & 5.53  &  1.92\\*
& npScarf\_wag & 5471159  &  2  &  5249462  &  0  &  6.98 &  1.72\\*
& npGraph-bwa & 5472856  &  2  &  5248714  & 0  & 8.06  &  1.41\\*
& npGraph-mm2 &  5472845 &  2  &  5248703  &  0 &  7.07 &  1.15\\*
& hybridSPAdes & 5472760  &  2  &  5248601  &  0  &  2.34 & 0.60\\*
& Unicycler & 5472697  & 2  & 5248545  &  0 & 2.41  &  0.35\\
\hline
\rowcolor{Gray}
\multicolumn{8}{l}{\emph{Mycobacterium tuberculosis} H37Rv} \\* %
\nobreakmidrule
\rowcolor{Gray}
& npScarf &  4446095 &  4  & 4389968   &  12  & 7.61  & 3.53\\*
\rowcolor{Gray}
& npScarf\_wag &  4407553 &  1  &  4407553  &  2  & 4.21  & 2.80\\*
\rowcolor{Gray}
& npGraph-bwa & 4411594  &  1  &  4411594  &  0  & 6.21  & 1.07\\*
\rowcolor{Gray}
& npGraph-mm2 & 4411387  & 1   &  4411387  &  0  & 6.28  & 0.73\\*
\rowcolor{Gray}
& hybridSPAdes & 4411162  &  1  &  4411162  &  0  &  1.41 &  0.32\\*
\rowcolor{Gray}
& Unicycler & 4411538  &  1  &  4411538  &  0  &  2.22 &  0.34\\
\hline
% \pagebreak
% \toprule
\multicolumn{8}{l}{\emph{Saccharomyces cerevisiae} S288c} \\* %
\nobreakmidrule
& npScarf &  11875932 &  20  &  896738  &  48  &  77.76 & 12.63 \\*
& npScarf\_wag & 12626448  &  37  &  630281  &  189  & 43.11  & 5.16 \\*
& npGraph-bwa & 11907663  &  37  &  774482  &  24 &  72.70 &  4.14\\*
& npGraph-mm2 & 11881030  &  39  &  795662  &  30 &  58.48 &  3.58\\*
& hybridSPAdes &  11968398 &  246  &  731962  &  4  & 12.20  & 0.89\\*
& Unicycler &  11847655 & 72  & 909114  &  0 & 21.81  &  1.04\\
\hline
\rowcolor{Gray}
\multicolumn{8}{l}{\emph{Shigella dysenteriae}  Sd197} \\* %
\nobreakmidrule
\rowcolor{Gray}
& npScarf &  4037570 &  70  & 179761   &  63  & 103.66  & 182.99\\*
\rowcolor{Gray}
& npScarf\_wag & 6571492  &  23  &  640874  &  1154  &  81.15 & 53.47\\*
\rowcolor{Gray}
& npGraph-bwa & 4574433  &  6  &  4382010  &  133  & 93.64  & 12.47\\*
\rowcolor{Gray}
& npGraph-mm2 & 4565129  &  10  &  1512588  &  193  & 86.63  & 13.29\\*
\rowcolor{Gray}
& hybridSPAdes &  4562642 & 244   &  77009  & 35   & 7.14  &  1.20\\*
\rowcolor{Gray}
& Unicycler & 4560901  & 3   &  4369231  &  0  & 11.88  & 1.05 \\
\hline
\multicolumn{8}{l}{\emph{Shigella sonnei} 53G} \\* %
\nobreakmidrule
& npScarf &  5335195 &  77  &  388533  &  68  & 104.26  &  116.07\\*
& npScarf\_wag & -  &  -  &  -  &  -  & -  &  -\\*
& npGraph-bwa & 5210154  &   4 &  4987138  & 14  & 59.84  &  1.02\\*
& npGraph-mm2 & 5212051  &  4  &   4989035 & 19  &  30.70 &  1.01\\*
& hybridSPAdes & 5223176  &  14  &  4987892  &  0  & 3.14  & 0.27\\*
& Unicycler &  5220517 &  5 &  4988548 & 0  & 7.39  &  0.52\\
\hline
\rowcolor{Gray}
\multicolumn{8}{l}{\emph{Streptococcus suis} BM407} \\* %
\nobreakmidrule
\rowcolor{Gray}
& npScarf &  2220434 & 4   &  2120037  &  9  & 48.28  & 48.92\\*
\rowcolor{Gray}
& npScarf\_wag & 2245214  &  4  &  2128290  &  3  & 85.79  & 3.85\\*
\rowcolor{Gray}
& npGraph-bwa & 2167238  &  6  &  2146682  &  0  & 26.08  & 0.69\\*
\rowcolor{Gray}
& npGraph-mm2 & 2166779  &  6  &  2146223  &  0  & 21.88  & 0.65\\*
\rowcolor{Gray}
& hybridSPAdes & 2147092  &  48  &  1437972  &  0  & 0.79  &  0.19\\*
\rowcolor{Gray}
& Unicycler &  2170829 &   2 &  2146250  &  0  &  2.67 & 0.32 \\
\hline
\end{longtable}

\begin{figure}[!hpt]
\centering
\subfloat[\emph{Citrobacter~freundii} CAV1374]{
	\includegraphics[width=.45\textwidth]{images/dp_c_freundii_cav1374.png}
}
\hfill
\subfloat[\emph{Klebsieall~oxytoca} CAV1015]{
	\includegraphics[width=.45\textwidth]{images/dp_k_oxytoca_cav1015.png}
}
\\
\subfloat[\emph{Enterobacter~cloacae} CAV1411]{
	\includegraphics[width=.45\textwidth]{images/dp_e_cloacae_cav1411.png}
}
\caption[Dotplot generated by MUMmer for assembly results of \unicycler{} versus \npgraph{}.]{Dotplot generated by MUMmer for assembly results of \unicycler{} versus \npgraph{}. Structural agreements between two methods were found in (a)~\emph{C.freundii} and (b)~\emph{K.oxytoca} assembly contigs. On the other hand, for (c)~\emph{E.cloacae} sample, there was a disagreement detected between 2 largest contigs given by two assembly algorithms. This case is investigated more thoroughly by using a reference from a same bacterial strain in Figure~\ref{supp_fig:npgraph_ref}.}
\label{supp_fig:npgraph_dotplot}
\end{figure}

\begin{figure}[!hpt]
\centering
\subfloat[\emph{E.~cloacae} \unicycler{} assembly versus reference genome]{
	\includegraphics[width=.5\textwidth]{images/dp_ref_unicycler.png}
}
\hfill
\subfloat[\emph{E.~cloacae} \npgraph{} assembly versus reference genome]{
	\includegraphics[width=.5\textwidth]{images/dp_ref_npgraph.png}
}
\caption[Alignments of a \emph{Enterobacter~cloacae} reference genome to assembly sequences generated by \unicycler{} and \npgraph{}]{Alignments of an \emph{Enterobacter~cloacae} reference genome to assembly sequences generated by  (a)~\unicycler{} and (b)~\npgraph{}. While the former presents a structural variant, the latter is virtually an 1-to-1 mapping.}
\label{supp_fig:npgraph_ref}
\end{figure}

\addtocontents{toc}{\protect\setcounter{tocdepth}{ 0}}
\chapter{Supplementary materials for Chapter 5}\label{app:concatemer}
%\addtocontents{toc}{\protect\setcounter{tocdepth}{ 0}}
%\addcontentsline{toc}{chapter}{Appendix \thechapter}
\newpage

\begin{figure}[!ht]
\centerline{\includegraphics[width=0.9\textwidth]{images/concat_count.jpg}}
\caption[Concatemer reads count]{Number of $k$-concatemers for each positive viral sample with barcode 08 (CaMV) and 09 (BSMYV). $k=1\dots 7$}
\label{supp_fig:concat_count}
\end{figure}

\begin{figure}[!ht]
\centering
\subfloat[ACF for a monomer read]{
	\includegraphics[width=.5\textwidth]{images/concatemer-1_std.png}
}
~
\subfloat[ACF for a 2-concatemers read]{
	\includegraphics[width=.5\textwidth]{images/concatemer-2_std.png}
}
\\
\subfloat[ACF for a 3-concatemers read]{
	\includegraphics[width=.5\textwidth]{images/concatemer-3_std.png}
}
~
\subfloat[ACF for a 6-concatemers read]{
	\includegraphics[width=.5\textwidth]{images/concatemer-6_std.png}
}
\\
\subfloat[ACF for the 7-concatemer read]{
	\includegraphics[width=.5\textwidth]{images/concatemer-7_std.png}
}
~
\subfloat[ACF for a random DNA sequence]{
	\includegraphics[width=.5\textwidth]{images/concatemer-random_std.png}
}

\caption{ACF values for a random synthetic read and several k-concatemers nanopore read detected in \emph{Cauliflower mosaic} sample (barcode 08).}
\label{supp_fig:concat_acf_dna}
\end{figure}

\begin{figure}[!hpt]
\centering
\subfloat[NSDF signal averaged by 10k window\label{supp_fig:acf_mpm10k}]{
	\includegraphics[width=.9\textwidth]{images/concatemer_mpm_10k.png}
}
\\
\subfloat[NSDF signal averaged by 20k window\label{supp_fig:acf_mpm20k}]{
	\includegraphics[width=.9\textwidth]{images/concatemer_mpm_20k.png}
}
\caption[NSDF signal processed after running average filter method with different window size]{NSDF signal processed by a running average filter step with different window size ($ws$): (\ref{supp_fig:acf_mpm10k}) $ws=10,000$ (\ref{supp_fig:acf_mpm20k}) $ws=20,000$. While the larger window sizes mean better de-noised signals, it also reduces the specificity of the local maxima. 
As can be seen from the Figure~\ref{supp_fig:acf_mpm20k}, the peaking regions are easier to spot but their blunt tips, making it more difficult to determine the exact coordinates. In Figure~\ref{supp_fig:acf_mpm10k} the window size $10,000$ returned clear spikes for easier monomer length evaluation. However, the seesaw pattern were not eliminated completely.}
\label{supp_fig:concat_acf_raw}
\end{figure}

\begin{figure}[!hpt]
\centering
\subfloat[NDSF for DNA sequence\label{supp_fig:mpm30_dna}]{
	\includegraphics[width=.9\textwidth]{images/nsdf-dna-cut30.png}
}
\\
\subfloat[NSDF for raw signal\label{supp_fig:mpm30_raw}]{
	\includegraphics[width=.9\textwidth]{images/nsdf-raw-cut30.png}
}
\caption[NSDF signal processed after smoothing with LPF at $cutFreq=30$]{7-concatemers NSDF signal processed after LPF using $cutFreq=30$ of: (\ref{supp_fig:mpm30_dna}) base-called DNA sequence (\ref{supp_fig:mpm30_raw}) raw signal sequence.}
\label{supp_fig:mpm30}
\end{figure}